PyMySQL Integration

lzchen opened the pull request:
Integration for [PyMySQL](https://pypi.org/project/PyMySQL/).
Fixes some documentation as well for other db integrations.
Leverages dbapi.

ocelotl commented on ext/opentelemetry-ext-dbapi/README.rst:
```diff
@@ -20,9 +20,9 @@ Usage
     trace.set_tracer_provider(TracerProvider())
     tracer = trace.get_tracer(__name__)
     # Ex: mysql.connector
-    trace_integration(tracer_provider(), mysql.connector, "connect", "mysql", "sql")
```
`tracer_provider` should not be imported anymore, also please check how `get_tracer` is being accessed.

ocelotl commented on ext/opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py:
```diff
@@ -125,6 +125,11 @@ def get_connection_attributes(self, connection):
         if self.database:
+            # PyMySQL encodes names with utf-8
+            try:
+                self.database = self.database.decode()
```
Just a suggestion, depending on the contents of `self.database` it may be a good idea to set the `errors` argument in `decode` to something else than the default [`strict`](https://docs.python.org/3/library/stdtypes.html#bytes.decode).

Also, if it is expected that some attribute won't be present, then it is better to handle this with `hasattr` instead of with a `try` that should be reserved for actually exceptional situations.

ocelotl commented on ext/opentelemetry-ext-pymysql/tests/test_pymysql_integration.py:
```diff
@@ -0,0 +1,43 @@
+            spec_set=True,
+            return_value=span,
+        )
+        start_as_current_span = start_current_span_patcher.start()
```
Maybe it is a good idea to add a call to `stop` at the end of the test case just for the sake of completeness?

hectorhdzg commented on ext/opentelemetry-ext-pymysql/tests/test_pymysql_integration.py:
```diff
@@ -0,0 +1,43 @@
+# Copyright 2019, OpenTelemetry Authors
```
```suggestion
# Copyright 2020, OpenTelemetry Authors
```

lzchen commented:
@hectorhdzg I've added functional tests, they are basically the exact same as Mysql except using the PyMySQL library

@ocelotl I believe I've addressed all your comments

mauriciovasquezbernal commented on docs/ext/pymysql/pymysql.rst:
```diff
@@ -0,0 +1,10 @@
+.. include:: ../../../ext/opentelemetry-ext-pymysql/README.rst
+
+
```
Currently we're not including that readme in these files anymore.

mauriciovasquezbernal commented on ext/opentelemetry-ext-mysql/setup.cfg:
```diff
@@ -41,8 +41,8 @@ package_dir=
 packages=find_namespace:
 install_requires =
     opentelemetry-api == 0.7.dev0
+    opentelemetry-ext-dbapi >= 0.7.dev0
```
```suggestion
    opentelemetry-ext-dbapi == 0.7.dev0
```

I don't remember the PR anymore but we decided to pin all the versions around for the time being.

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/setup.cfg:
```diff
@@ -0,0 +1,47 @@
+# Copyright 2020, OpenTelemetry Authors
```
```suggestion
# Copyright The OpenTelemetry Authors
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/setup.cfg:
```diff
@@ -0,0 +1,47 @@
+long_description_content_type = text/x-rst
+author = OpenTelemetry Authors
+author_email = cncf-opentelemetry-contributors@lists.cncf.io
+url = https://github.com/open-telemetry/opentelemetry-python/ext/opentelemetry-ext-pymysql
```
```suggestion
url = https://github.com/open-telemetry/opentelemetry-python/tree/master/ext/opentelemetry-ext-pymysql
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/setup.cfg:
```diff
@@ -0,0 +1,47 @@
+    Programming Language :: Python :: 3.4
+    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.7
```
```suggestion
    Programming Language :: Python :: 3.7
    Programming Language :: Python :: 3.8
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/setup.cfg:
```diff
@@ -0,0 +1,47 @@
+    =src
+packages=find_namespace:
+install_requires =
+    opentelemetry-api >= 0.6.dev0
+    opentelemetry-ext-dbapi >= 0.6.dev0
```
```suggestion
    opentelemetry-api == 0.7.dev0
    opentelemetry-ext-dbapi == 0.7.dev0
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/setup.py:
```diff
@@ -0,0 +1,26 @@
+# Copyright 2020, OpenTelemetry Authors
```
```suggestion
# Copyright The OpenTelemetry Authors
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/src/opentelemetry/ext/pymysql/__init__.py:
```diff
@@ -0,0 +1,63 @@
+    trace.set_tracer_provider(TracerProvider())
+    tracer = trace.get_tracer(__name__)
+    trace_integration(tracer)
+    cnx = pymysql.connect(database='MySQL_Database')
```
```suggestion
    cnx = pymysql.connect(database="MySQL_Database")
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-pymysql/tests/test_pymysql_integration.py:
```diff
@@ -0,0 +1,45 @@
+    def test_trace_integration(self):
+        tracer = trace_api.DefaultTracer()
+        span = mock.create_autospec(trace_api.Span, spec_set=True)
+        start_current_span_patcher = mock.patch.object(
```
Maybe you want to use `TestBase` as done here https://github.com/open-telemetry/opentelemetry-python/pull/586?

mauriciovasquezbernal commented on ext/opentelemetry-ext-docker-tests/tests/pymysql/test_pymysql_functional.py:
```diff
@@ -0,0 +1,108 @@
+MYSQL_DB_NAME = os.getenv("MYSQL_DB_NAME ", "opentelemetry-tests")
+
+
+class TestFunctionalPyMysql(unittest.TestCase):
```
Also here, using TestBase could be a lot.

mauriciovasquezbernal commented on ext/opentelemetry-ext-docker-tests/tests/pymysql/test_pymysql_functional.py:
```diff
@@ -0,0 +1,108 @@
+    def test_callproc(self):
+        """Should create a child span for callproc
+        """
+        with self._tracer.start_as_current_span("rootSpan"), self.assertRaises(
+            Exception
+        ):
```
I think two separated `with` statements would be a lot clearer in this particular case.
